anomalib/data/base/dataset.py

```diff
@@ -0,0 +1,155 @@
+"""Anomalib dataset base class."""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import copy
+import logging
+from abc import ABC, abstractmethod
+from typing import Dict, Sequence, Union
+
+import cv2
+import numpy as np
+import pandas as pd
+from pandas import DataFrame
+from torch import Tensor
+from torch.utils.data import Dataset
+
+from anomalib.data.utils import read_image
+from anomalib.pre_processing import PreProcessor
+
+logger = logging.getLogger(__name__)
+
+
+class AnomalibDataset(Dataset, ABC):
+    """Anomalib dataset."""
+
+    def __init__(self, task: str, pre_process: PreProcessor):
+        super().__init__()
+        self.task = task
+        self.pre_process = pre_process
+        self._samples = None
+
+    def __len__(self) -> int:
+        """Get length of the dataset."""
+        assert isinstance(self._samples, DataFrame)
+        return len(self._samples)
+
+    def subsample(self, indices: Sequence[int], inplace=False) -> AnomalibDataset:
+        """Subsamples the dataset at the provided indices.
+
+        Args:
+            indices (Sequence[int]): Indices at which the dataset is to be subsampled.
+            inplace (bool): When true, the subsampling will be performed on the instance itself.
+        """
+        dataset = self if inplace else copy.deepcopy(self)
+        dataset.samples = self.samples.iloc[indices].reset_index(drop=True)
+        return dataset
+
+    @property
+    def is_setup(self) -> bool:
+        """Checks if setup() been called."""
+        return isinstance(self._samples, DataFrame)
+
+    @property
+    def samples(self) -> DataFrame:
+        """Get the samples dataframe."""
+        if not self.is_setup:
+            raise RuntimeError("Dataset is not setup yet. Call setup() first.")
+        return self._samples
+
+    @samples.setter
+    def samples(self, samples: DataFrame):
+        """Overwrite the samples with a new dataframe.
+
+        Args:
+            samples (DataFrame): DataFrame with new samples.
+        """
+        self._samples = samples.sort_values(by="image_path", ignore_index=True)
```

Comment:
the class depends a lot on `samples` so it would be easier to debug some basic expectations about it

```suggestion
        assert isinstance(samples, DataFrame), f"samples must be a pandas.DataFrame, found {type(samples)}"
        EXPECTED_COLUMNS =  _EXPECTED_COLS_PERTASK[self.task] 
        assert all(col in samples.columns for col in EXPECTED_COLUMNS), f"samples must have (at least) columns {EXPECTED_COLUMNS}, found {samples.columns}"
        assert samples["image_path"].apply(lambda p: Path(p).exists()).all(), "missing file pointed by samples"
        
        self._samples = samples.sort_values(by="image_path", ignore_index=True)
```

these problems are easier to correct when the show up at setup rather then training phase

---

this should be up there in the module level

```python
_EXPECTED_COlS_CLASSIFICATION = ["image_path", "label", "label_index", "mask_path", "split"]
_EXPECTED_COLS_SEGMENTATION = _EXPECTED_COlS_CLASSIFICATION + ["mask_path"]
_EXPECTED_COLS_PERTASK = {
    "classification": _EXPECTED_COlS_CLASSIFICATION,
    "segmentaion": _EXPECTED_COLS_SEGMENTATION,
}
```

Reply:
Thanks, great suggestion!